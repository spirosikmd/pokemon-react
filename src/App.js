import React, {Component} from 'react';
import {getPokemons, getPokemon} from './api';
import Pokemon from './Pokemon';
import PokemonList from './PokemonList';
import './App.css';

class App extends Component {
  constructor(props) {
    super(props);

    this.state = {
      pokemons: [],
      selectedPokemon: null,
      isLoading: true,
      previousLink: null,
      nextLink: null
    };

    this.handlePreviousClick = this.handlePreviousClick.bind(this);
    this.handleNextClick = this.handleNextClick.bind(this);
    this.handlePokemonSelect = this.handlePokemonSelect.bind(this);
  }

  componentDidMount() {
    getPokemons()
      .then(this.onGetPokemonsSuccess.bind(this));
  }

  handlePokemonSelect(pokemonUrl) {
    getPokemon(pokemonUrl)
      .then((pokemonResponse) => {
        const selectedPokemon = Object.assign({}, pokemonResponse);
        this.setState({selectedPokemon});
      });
  }

  onGetPokemonsSuccess(pokemonsResponse) {
    const pokemons = [...pokemonsResponse.results];
    this.setState({
      pokemons,
      isLoading: false,
      nextLink: pokemonsResponse.next,
      previousLink: pokemonsResponse.previous
    });
  }

  handlePreviousClick() {
    const previousLink = this.state.previousLink;
    getPokemons(previousLink)
      .then(this.onGetPokemonsSuccess.bind(this));
  }

  handleNextClick() {
    const nextLink = this.state.nextLink;
    getPokemons(nextLink)
      .then(this.onGetPokemonsSuccess.bind(this));
  }

  render() {
    const isLoading = this.state.isLoading;
    const pokemons = this.state.pokemons;
    const selectedPokemon = this.state.selectedPokemon;
    const previousLink = this.state.previousLink;
    const nextLink = this.state.nextLink;

    return (
      <div>
        {!isLoading ? (
          <div>
            <PokemonList pokemons={pokemons} onPokemonSelect={this.handlePokemonSelect} />
            <button onClick={this.handlePreviousClick} disabled={!previousLink}>previous</button>
            <button onClick={this.handleNextClick} disabled={!nextLink}>next</button>
          </div>
        ) : (
          <div>Loading...</div>
        )}
        {selectedPokemon ? (
          <Pokemon pokemon={selectedPokemon} />
        ): (
          <div>Select a pokemon!</div>
        )}
      </div>
    );
  }
}

export default App;
