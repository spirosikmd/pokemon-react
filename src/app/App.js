import React, { PureComponent } from 'react';
import { getSinglePokemon, getAllPokemon } from '../api';
import PokemonDetail from '../pokemon-detail/PokemonDetail';
import PokemonList from '../pokemon-list/PokemonList';
import Filters from '../filters/Filters';
import './App.css';
import { catchPokemon, getMyPokemon, removePokemon } from '../db';

class App extends PureComponent {
  constructor(props) {
    super(props);

    this.state = {
      allPokemon: [],
      myPokemon: [],
      selectedPokemon: null,
      isLoading: true,
      showPokemon: false,
      count: null,
      searchText: '',
      myPokemonOnly: false,
      perPage: 20,
      selectedPage: 0,
    };

    this.handlePokemonClick = this.handlePokemonClick.bind(this);
    this.handleCloseModal = this.handleCloseModal.bind(this);
    this.handleCatchClick = this.handleCatchClick.bind(this);
    this.handleSearchTextInput = this.handleSearchTextInput.bind(this);
    this.handleMyPokemonInput = this.handleMyPokemonInput.bind(this);
    this.handleFiltersReset = this.handleFiltersReset.bind(this);
    this.handlePageChange = this.handlePageChange.bind(this);
    this.handlePerPageChange = this.handlePerPageChange.bind(this);
  }

  componentDidMount() {
    getAllPokemon().then(this.onGetAllPokemonSuccess.bind(this));
    this.updateMyPokemon();
  }

  updateMyPokemon() {
    return getMyPokemon().then(myPokemon => {
      this.setState({ myPokemon: [...myPokemon] });
    });
  }

  filterPokemon(allPokemon, myPokemon, searchText, myPokemonOnly) {
    return allPokemon
      .filter(pokemon => {
        return !searchText || pokemon.name.indexOf(searchText) !== -1;
      })
      .filter(pokemon => {
        if (!myPokemonOnly) {
          return true;
        }
        return this.isMyPokemon(myPokemon, pokemon.name);
      });
  }

  paginatePokemon(filteredPokemon, perPage, selectedPage) {
    const start = Math.ceil(selectedPage * perPage);
    const end = start + perPage - 1;
    return filteredPokemon.slice(start, end + 1);
  }

  handleSearchTextInput(searchText) {
    this.setState({ searchText, selectedPage: 0 });
  }

  handleMyPokemonInput(myPokemonOnly) {
    this.setState({ myPokemonOnly, selectedPage: 0 });
  }

  handlePokemonClick(pokemonName) {
    getSinglePokemon(pokemonName).then(pokemonResponse => {
      const selectedPokemon = Object.assign({}, pokemonResponse);
      this.setState({ selectedPokemon, showPokemon: true });
    });
  }

  handleCatchClick(myPokemon, pokemonNameToCatch) {
    if (!this.isMyPokemon(myPokemon, pokemonNameToCatch)) {
      catchPokemon(pokemonNameToCatch).then(this.updateMyPokemon.bind(this));
      return;
    }

    removePokemon(pokemonNameToCatch).then(this.updateMyPokemon.bind(this));
  }

  handleFiltersReset() {
    this.setState({ searchText: '', myPokemonOnly: false, selectedPage: 0 });
  }

  handlePageChange({ selected }) {
    this.setState({ selectedPage: selected });
  }

  handleCloseModal() {
    this.setState({ showPokemon: false });
  }

  handlePerPageChange(perPage) {
    this.setState({ perPage });
  }

  isMyPokemon(myPokemon, pokemonName) {
    return (
      myPokemon.find(pokemon => pokemon.name === pokemonName) !== undefined
    );
  }

  onGetAllPokemonSuccess({ results, count }) {
    const allPokemon = [...results];
    this.setState({
      allPokemon,
      isLoading: false,
      count,
    });
  }

  getPageCount(filteredPokemonLength, perPage) {
    return Math.ceil(filteredPokemonLength / perPage);
  }

  render() {
    const {
      selectedPokemon,
      count,
      myPokemon,
      searchText,
      myPokemonOnly,
      perPage,
    } = this.state;

    const filteredPokemon = this.filterPokemon(
      this.state.allPokemon,
      myPokemon,
      searchText,
      myPokemonOnly
    );
    const paginatedPokemon = this.paginatePokemon(
      filteredPokemon,
      perPage,
      this.state.selectedPage
    );

    const pageCount = this.getPageCount(filteredPokemon.length, perPage);

    return (
      <div>
        {count && <div>In total there are {count} pokemon!</div>}

        <Filters
          searchText={searchText}
          myPokemonOnly={myPokemonOnly}
          onSearchTextInput={this.handleSearchTextInput}
          onMyPokemonInput={this.handleMyPokemonInput}
        />

        {!this.state.isLoading
          ? <div>
              <PokemonList
                allPokemon={paginatedPokemon}
                myPokemon={myPokemon}
                onPokemonClick={this.handlePokemonClick}
                onCatchClick={name => this.handleCatchClick(myPokemon, name)}
                onFiltersReset={this.handleFiltersReset}
                onPageChange={this.handlePageChange}
                onPerPageChange={this.handlePerPageChange}
                pageCount={pageCount}
                perPage={perPage}
              />
            </div>
          : <div>Loading...</div>}

        {selectedPokemon &&
          <PokemonDetail
            pokemon={selectedPokemon}
            show={this.state.showPokemon}
            onCloseModal={this.handleCloseModal}
          />}
      </div>
    );
  }
}

export default App;
