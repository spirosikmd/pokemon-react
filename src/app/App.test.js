import React from 'react';
import renderer from 'react-test-renderer';
import App from './App';

describe('App', () => {
  let app;

  beforeEach(() => {
    app = new App();
    app.setState = jest.fn();
  });

  afterEach(() => {
    app = null;
  });

  test('renders', () => {
    const component = renderer.create(<App />);
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('pageCount returns the page count', () => {
    expect(app.getPageCount(100, 15)).toBe(7);
  });

  test('onGetPokemonsSuccess sets pokemons, isLoading, and count in state', () => {
    app.onGetAllPokemonSuccess({
      results: [
        {
          name: 'pikachu',
        },
      ],
      count: 1,
    });
    expect(app.setState.mock.calls[0][0]).toEqual({
      allPokemon: [
        {
          name: 'pikachu',
        },
      ],
      isLoading: false,
      count: 1,
    });
  });

  describe('isMyPokemon', () => {
    let myPokemon;

    beforeEach(() => {
      myPokemon = [
        {
          name: 'pikachu',
        },
      ];
    });

    afterEach(() => {
      myPokemon = null;
    });

    test('returns true when pokemon name is my pokemon', () => {
      const result = app.isMyPokemon(myPokemon, 'pikachu');
      expect(result).toBe(true);
    });

    test('returns false when pokemon name is not my pokemon', () => {
      const result = app.isMyPokemon(myPokemon, 'bulbasar');
      expect(result).toBe(false);
    });
  });

  test('handleCloseModal sets showPokemon to false', () => {
    app.handleCloseModal();
    expect(app.setState.mock.calls[0][0]).toEqual({
      showPokemon: false,
    });
  });

  test('handlePageChange sets selectedPage to given selected page', () => {
    app.handlePageChange({ selected: 2 });
    expect(app.setState.mock.calls[0][0]).toEqual({
      selectedPage: 2,
    });
  });

  test('handleFiltersReset resets searchText, myPokemonOnly, and selectedPage', () => {
    app.handleFiltersReset();
    expect(app.setState.mock.calls[0][0]).toEqual({
      searchText: '',
      myPokemonOnly: false,
      selectedPage: 0,
    });
  });

  test('handleMyPokemonInput sets the new myPokemonOnly and resets selectedPage', () => {
    app.handleMyPokemonInput(true);
    expect(app.setState.mock.calls[0][0]).toEqual({
      myPokemonOnly: true,
      selectedPage: 0,
    });
  });

  test('handleSearchTextInput sets the new searchText and resets selectedPage', () => {
    app.handleSearchTextInput('pikachu');
    expect(app.setState.mock.calls[0][0]).toEqual({
      searchText: 'pikachu',
      selectedPage: 0,
    });
  });

  test('paginatePokemons returns the paginated pokemons', () => {
    const filteredPokemons = [
      {
        name: 'pikachu',
      },
      {
        name: 'bulbasar',
      },
      {
        name: 'charmeleon',
      },
      {
        name: 'metapod',
      },
      {
        name: 'rattata',
      },
    ];
    const paginated = app.paginatePokemon(filteredPokemons, 2, 1);
    expect(paginated.length).toBe(2);
    expect(paginated[0].name).toBe('charmeleon');
    expect(paginated[1].name).toBe('metapod');
  });

  test('filterPokemons filters pokemons based on searchText and myPokemonOnly', () => {
    const pokemons = [
      {
        name: 'pikachu',
      },
      {
        name: 'bulbasar',
      },
      {
        name: 'charmeleon',
      },
      {
        name: 'metapod',
      },
      {
        name: 'rattata',
      },
    ];
    const myPokemon = [
      {
        name: 'metapod',
      },
    ];

    let filtered = app.filterPokemon(pokemons, myPokemon, 'pika', false);
    expect(filtered.length).toBe(1);
    expect(filtered[0].name).toBe('pikachu');

    filtered = app.filterPokemon(pokemons, myPokemon, 'e', true);
    expect(filtered.length).toBe(1);
    expect(filtered[0].name).toBe('metapod');
  });

  test('handlePerPageChange sets perPage to state', () => {
    app.handlePerPageChange(40);
    expect(app.setState.mock.calls.length).toBe(1);
    expect(app.setState.mock.calls[0][0]).toEqual({
      perPage: 40,
    });
  });
});
