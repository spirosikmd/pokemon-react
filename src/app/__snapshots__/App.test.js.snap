exports[`App renders 1`] = `
<div>
  <form>
    <label>
      <input
        onChange={[Function]}
        placeholder="Search..."
        type="text"
        value="" />
    </label>
    <p>
      <label>
        <input
          checked={false}
          onChange={[Function]}
          type="checkbox" />
        Only show my pokemon
      </label>
    </p>
  </form>
  <div>
    Loading...
  </div>
</div>
`;
